packages/fxa-settings/src/lib/gql-key-stretch-upgrade.ts

```diff
@@ -0,0 +1,263 @@
+        sensitiveDataClient.KeyStretchUpgradeData.email,
+        sensitiveDataClient.KeyStretchUpgradeData.v1Credentials,
+        sensitiveDataClient.KeyStretchUpgradeData.v2Credentials,
+        sessionId
+      );
+      return true;
+    }
+  } catch (error) {
+    // NO-OP Don't let a key stretching upgrade issue prevent sign in.
+    // Note that the upgradeClient reports errors to sentry.
+  } finally {
+    // Clear out the state. No reason to keep trying this...
+    // Note that the upgradeClient will report issues to Sentry.
+    sensitiveDataClient.KeyStretchUpgradeData = undefined;
+  }
+  return false;
+}
+
+/**
+ * Handles upgrade process for key stretching
+ */
+export class GqlKeyStretchUpgrade {
+  constructor(
+    private readonly stage: string,
+    private readonly gqlCredentialStatus: MutationFunction<CredentialStatusResponse>,
+    private readonly gqlGetWrappedKeys: MutationFunction<GetAccountKeysResponse>,
+    private readonly gqlPasswordChangeStart: MutationFunction<PasswordChangeStartResponse>,
+    private readonly gqlPasswordChangeFinish: MutationFunction<PasswordChangeFinishResponse>
+  ) {}
+
+  async getCredentials(
+    email: string,
+    password: string,
+    v2Enabled: boolean
+  ): Promise<{
+    v1Credentials: V1Credentials;
+    v2Credentials?: V2Credentials;
+    credentialStatus?: CredentialStatus;
+  }> {
+    const v1Credentials = await getCredentials(email, password);
+    if (v2Enabled) {
+      const credentialStatus = await this.getCredentialsStatus(email);
+      if (credentialStatus) {
+        const v2Credentials = await getCredentialsV2({
+          password,
+          clientSalt: credentialStatus?.clientSalt || createSaltV2(),
+        });
+        return {
+          credentialStatus,
+          v1Credentials,
+          v2Credentials,
+        };
+      }
+    }
+    return {
+      v1Credentials,
+    };
+  }
+
+  async upgrade(
+    email: string,
+    v1Credentials: V1Credentials,
+    v2Credentials: V2Credentials,
+    sessionToken: string
+  ): Promise<boolean> {
+    let result1 = await this.startUpgrade(email, v1Credentials);
+
+    if (result1?.keyFetchToken && result1?.passwordChangeToken) {
+      const result2 = await this.getWrappedKeys(result1.keyFetchToken);
+
+      if (result2?.wrapKB) {
+        await this.finishUpgrade(
+          result2?.wrapKB,
+          result1.passwordChangeToken,
+          v1Credentials,
+          v2Credentials,
+          sessionToken
+        );
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private async getCredentialsStatus(
+    email: string
+  ): Promise<CredentialStatus | undefined> {
+    try {
+      const result = await this.gqlCredentialStatus({
+        variables: {
+          input: email,
+        },
+      });
+      return result.data?.credentialStatus;
+    } catch (error) {
+      Sentry.captureMessage(
+        `Failure to finish v2 upgrade. Could not get credential status during ${this.stage}`,
```

Comment:
nit: maybe mention key stretching in the message?  Same with the Sentry errors below.